src/main/java/com/google/devtools/build/lib/exec/SpawnLogReconstructor.java

```diff
@@ -212,6 +225,136 @@ private static File reconstructSymlink(ExecLogEntry.UnresolvedSymlink entry) {
         .build();
   }
 
+  private Pair<String, Collection<File>> reconstructRunfilesDir(
+      ExecLogEntry.RunfilesTree runfilesTree) throws IOException {
+    var artifacts = reconstructInputs(runfilesTree.getArtifactsId());
+    LinkedHashMap<String, File> builder =
+        new LinkedHashMap<>(runfilesTree.getSymlinksCount() + artifacts.size());
+    for (var symlink : runfilesTree.getSymlinksMap().entrySet()) {
+      String newPath = runfilesTree.getPath() + "/" + symlink.getKey();
+      for (var file : reconstructRunfilesSymlinkTarget(newPath, symlink.getValue())) {
+        builder.put(newPath, file);
+      }
+    }
+    Collection<File> inputs = reconstructInputs(runfilesTree.getArtifactsId()).values();
+    inputs.stream()
+        .flatMap(
+            file ->
+                getRunfilesPaths(file.getPath(), runfilesTree.getLegacyExternalRunfiles())
+                    .map(
+                        relativePath ->
+                            file.toBuilder()
+                                .setPath(runfilesTree.getPath() + "/" + relativePath)
+                                .build()))
+        .forEach(file -> builder.put(file.getPath(), file));
+    getDotRunfileFileIfNeeded(runfilesTree).ifPresent(file -> builder.put(file.getPath(), file));
```

Comment:
We could avoid a second traversal here by keeping track of a boolean during the first traversal, but up to you; we can defer until benchmarks prove it worthwhile.